### ISSUE TITLE
MLIL SSA form: constant folding breaks SSA variable dependence

### ISSUE THREAD

**Issue by DefCon42** [Vector35/binaryninja-api#1481]
NOTE: I could be doing something wrong in my dead variable elimination pass, but just in case I'm going to post an issue here.

When a constant is folded into an MLIL SSA instruction, the instruction that sets that constant is considered dead. For example:

![image](https://user-images.githubusercontent.com/5364774/68997244-3fecd100-0872-11ea-996f-14abef1e78aa.png)
![image](https://user-images.githubusercontent.com/5364774/68997251-49763900-0872-11ea-80bf-d74c44eac5eb.png)

After I run this simple dead variable elimination pass:
```
for ssa_instruction in ssa_form.instructions:
  if ssa_instruction.address not in ssa_set:
    ssa_set[ssa_instruction.address] = []
  ssa_set[ssa_instruction.address].append(ssa_instruction)

for address, ssa_instructions in ssa_set.items():
  eliminate = True
  for ssa_instruction in ssa_instructions:
    if not eliminate:
      break
    for vw in ssa_instruction.vars_written:
      if ssa_form.is_ssa_var_live(vw):
        eliminate = False
        break
  if eliminate:
    bv.convert_to_nop(address)
```

R13 gets set to an undetermined value because the instruction setting it to a constant gets removed, and as such most of the function breaks.

![image](https://user-images.githubusercontent.com/5364774/68997253-509d4700-0872-11ea-994b-5d1b7795237a.png)

**Comment by joshwatson** [comment]
Your urls don’t work. You can paste images directly into the issue text box

**Comment by DefCon42** [comment]
> 
> 
> Your urls don’t work. You can paste images directly into the issue text box

Updated, sorry about that.

**Comment by joshwatson** [comment]
MLIL optimizes away constants, you should not rely on it to remove dead variables. Use LLIL or mapped MLIL for that

**Comment by DefCon42** [comment]
Gotcha.

**Comment by rollsafe** [comment]
This is not a bug. This is because the MLIL is generated based on the assembly. In this case trying to use the MLIL dead-code elimination results on the assembly is the source of the incorrectness: MLIL and assembly don't correspond one-to-one. That's why Josh suggested using LLIL instead (where the correspondence is much stronger).

Here is a more detailed explanation. Suppose you have code `r13 = 0; rcx = rcx + r13`. The constant `0` will be propagated to `r13 = 0; rcx = rcx + 0`. It would appear here that `r13 = 0` is redundant, so we want to eliminate this assignment. However if we NOP out the corresponding assembly instruction, there is still a def-use chain from the def `r13 = 0` to the use `rcx + r13`, **even though the constant has been propagated**. The propagated value comes from that initial assignment that now seems dead. As a result, r13 value becomes unknown, even though it was folded before.

**Comment by rollsafe** [comment]
Reopening this pending further investigation for now

**Comment by bpotchik** [comment]
This should now be fixed in 1.3.2088-dev. Build should be out shortly.
